### PR TITLE
Refractor progress tracker

### PR DESCRIPTION
**Pull request by JingwenGu0829**
Here is the trackboard for the refractor progress. 

## Overall Design

Mostly follow from https://github.com/sgl-project/sglang-omni/issues/188#issuecomment-4160130935

- [ ] Finalized design doc publish 
- [ ] Quickstart doc for new model support

##  Code Quality

- [ ]  AI slop clearup @JingwenGu0829 : silent fallback, boilerplate code, over-chatty comment...
Feel free to comment on this pr directly, or dm @JingwenGu0829 , or commit directly, if you are found any notable AI slop.

- [ ] Code quality / Design review @yxs (Potentially @FrankLeeeee @zhaochenyang20 )

##  Model accuracy/performance

- [ ] Benchmark Qwen-Omni for correctness and speed
              - [ ] Debug if there's correctness/speed issue
- [ ] Benchmark Fish Audio for correctness and speed
              - [ ]  Debug if there's correctness/speed issue

## In-progress features

- [ ] Ming-Omni support @yuan-luo 
- [ ] Flow Matching/Diffusion @FrankLeeeee 
- [ ] Streaming real-time input @PopSoda2002 
- [ ] TP Support @edwingao28 @yuan-luo 

- [ ] Same GPU, multiple stage's memory management 
- [ ] Server args config pipeline (server CLI support) @JingwenGu0829 

## Optimizatoins 

Qwen-Omni:
- [ ] Piecewise CUDA Graph for talker, high performance/precision talker backend @JingwenGu0829 
- [ ] High-performance code predictor backend/ KV cache management (half finished by @JingwenGu0829, need help)
- [ ]

### CODE REVIEW

**Comment by yuan-luo** [comment]
Could you create a design spec for this refactor? It would be more convenient to understand your design in a whole picture.

**Comment by edwingao28** [comment]
Direction looks good @JingwenGu0829. Some comments for TP: since TP is still the open part, the key question before going further is are we still building TP as single logical stage + leader-owned control/scheduler + follower-only compute?

If yes, config surface / process topology / streaming / failure handling all stay much simpler.
If not, each `tp_rank` risks behaving like a full stage and complexity grows quickly.

Given the current PR direction and existing TP implementation, I think we should first pin down:

1. TP ownership  
One logical stage has a single external control plane. Coordinator talks only to the stage leader; rank 0 owns scheduler/control; followers only participate in model forward / collectives.

2. Request distribution / lockstep  
Leader holds the single scheduler and broadcasts requests + control to followers. Followers are compute-only. This keeps all ranks in the same request order and collective step order.

3. Single source of truth for TP params  
`StageConfig` expresses logical intent only; `mp_runner` resolves per-rank runtime spec; `ModelWorker` / sglang receive fully resolved `tp_size`, `tp_rank`, GPU placement, `nccl_port`, and `server_args.tp_size`, with no inference / defaulting split across layers.

Pinning these semantics first could make the implementation path much clearer and keep the refactor aligned with the current TP model.

**Comment by JingwenGu0829** [comment]
Hi @edwingao28 Maybe you can take a look at the TP described in the design doc https://github.com/sgl-project/sglang-omni/issues/188#issuecomment-4160130935. So basically everything in TP stays the same as sglang upstream.  If a stage's model requires tp(we assign that in config), then the stage spawns multiple processes, each with a scheduler+modelrunner. The rank0 communicate with the stage directly. 

To detaildly answer your question I let AI framed the phrasing, but it still aligns with above:

```
Yes, that is the intended TP semantics.

  TP in `sglang-omni` is meant to stay aligned with upstream `sglang`: one logical stage, one external
  control plane, and the Coordinator/Stage talks only to TP rank 0. Non-zero TP ranks are not independent
  stages.

  The only nuance is that, like upstream `sglang`, each TP rank still runs a local scheduler + model worker
  so all ranks can execute the same schedule in lockstep and participate in native NCCL collectives. But
  those follower schedulers are internal replicas, not separate control owners.

  Concretely: rank 0 receives requests/control from the stage layer, broadcasts them to peer ranks, all
  ranks run the same scheduling / forward steps, and only rank 0 sends results downstream.

  So I agree with the three semantics you listed:
  1. `StageConfig` should stay logical-only.
  2. `mp_runner` should resolve per-rank runtime state (`tp_rank`, `tp_size`, GPU placement, `nccl_port`,
  `server_args.tp_size`).
  3. `ModelWorker` / `sglang` should receive fully resolved TP params, with no split inference/defaulting
  across layers.
```

**Comment by JingwenGu0829** [comment]
Hi @yuan-luo I think most of the pipeline aligns with the original design spec https://github.com/sgl-project/sglang-omni/issues/188#issuecomment-4160130935 

Things mostly stay the same, except I found the modelrunner of fish and qwen-omni don't align that well to be abstracted into 1 modelrunner above base. I will update the doc on this part after the design decision is condensed and the pipeline is free with bugs. But mostly I think it does not effect the developement of Ming. Thanks!

File: sglang_omni/config/__init__.py [hunk 1]
```diff
@@ -1,23 +1,16 @@
 # SPDX-License-Identifier: Apache-2.0
 from sglang_omni.config.compiler import compile_pipeline
-from sglang_omni.config.runner import PipelineRunner, build_pipeline_runner
 from sglang_omni.config.schema import (
     EndpointsConfig,
-    ExecutorConfig,
```
Comment (yxs) [inline sglang_omni/config/__init__.py:6]: Removing these also drops them from `sglang_omni.config`'s public surface, but `models/ming_omni/` and `models/voxtral_tts` still import `ExecutorConfig` and `InputHandlerConfig`, `sgl-omni serve` will fail.

port ming/voxtral in this PR, or drop the two packages until their follow-up PRs.

File: sglang_omni/scheduling/omni_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+    Composition strategy:
+        SGLang scheduling methods (``get_next_batch_to_run``,
+        ``process_batch_result``, …) are looked up on the upstream
+        ``Scheduler`` *class* via ``__getattr__`` and called with this
+        instance as ``self``.  Methods we override (``recv_requests``,
+        ``process_input_requests``, ``run_batch``, ``send_to_tokenizer``)
+        are defined directly on this class and take precedence.
+    """
+
+    def __init__(
```
Comment (yxs) [inline sglang_omni/scheduling/omni_scheduler.py:83]: Neither copy nor inherit, borrowing upstream's batch-selection methods via unbound-method binding + manually mirroring the state they read is pragmatic.

Better to comment on `__init__` and `__getattr__` before upstream factoring Scheduler into a clean ScheduleCore + server shell so we can actually inherit.
1. names the two coupling surfaces (method reflection + state mirror)
2. pins the upstream SGLang version this was designed against

File: sglang_omni/scheduling/simple_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            while self._running:
+                msg = self._next_message()
+                if msg is None:
+                    continue
+
+                if msg.type == "new_request":
+                    try:
+                        batch = self._collect_batch(msg)
+                        self._run_batch(batch, loop)
+                    except Exception:
```
Comment (yxs) [inline sglang_omni/scheduling/simple_scheduler.py]: `_run_batch` exception is logged but never emitted to `self.outbox`, so `Stage._drain_outbox` sees nothing and Coordinator's `submit()` future hangs until client-side timeout.

Violates rule 1 here: https://github.com/sgl-project/sglang-omni/issues/188#issuecomment-4262497139

Fix: emit an error-tagged `OutgoingMessage` (new `type="error"` or reuse `type="result"` with an error field) so Stage can route to `_send_failure`.

File: sglang_omni/pipeline/stage/runtime.py [hunk 4]
```diff
@@ -88,385 +83,205 @@ def __init__(
...
+                        import torch
+
+                        torch.cuda.set_device(int(self.gpu_id))
+                        logger.info(
+                            "Scheduler thread for stage %s set CUDA device to %s",
+                            self.name,
+                            self.gpu_id,
+                        )
+                    self.scheduler.start()
+                except Exception:
```
Comment (yxs) [inline sglang_omni/pipeline/stage/runtime.py]: Scheduler thread crash → `_running = False` only, in-flight requests hang until timeout.

pattern 1: https://github.com/sgl-project/sglang-omni/issues/188#issuecomment-4262497139

File: sglang_omni/pipeline/stage/runtime.py [hunk 5]
```diff
@@ -88,385 +83,205 @@ def __init__(
...
-            self.router.enqueue(work)
-            # Flush pending stream data that arrived before this request was assigned
-            pending_stream = self._pending_stream_data.pop(request_id, [])
-            for pending in pending_stream:
+        # Read payload from relay
+        try:
+            payload = await relay_io.read_payload(
+                self.relay, request_id, msg.shm_metadata
+            )
+        except Exception:
```
Comment (yxs) [inline sglang_omni/pipeline/stage/runtime.py]: same, pattern 1: #188 (comment)

File: sglang_omni/pipeline/stage/runtime.py (hunk 4 above)
Comment (JingwenGu0829) [reply sglang_omni/pipeline/stage/runtime.py]: yeah these are horrible AI slops, will fix these!

File: sglang_omni/scheduling/simple_scheduler.py (hunk 3 above)
Comment (JingwenGu0829) [reply sglang_omni/scheduling/simple_scheduler.py]: AI slops again. Will fix! Thanks!

File: sglang_omni/scheduling/omni_scheduler.py (hunk 2 above)
Comment (JingwenGu0829) [reply sglang_omni/scheduling/omni_scheduler.py:83]: Yeah this is a harsh design decision yet. Right now we are following the same composition pattern as sglang-diffusion by design. Will add comment on it.

File: sglang_omni/config/__init__.py (hunk 1 above)
Comment (JingwenGu0829) [reply sglang_omni/config/__init__.py:6]: @yuan-luo Do you think we just implement Ming in this refractor directly, or do follow-up PR here?

**Comment by zhaochenyang20** [comment]
CI management during the refactor will be painful. My proposal: don't modify files in the current pipeline — instead, open a new folder  sglang-omni/v1 , and add a flag  --use-v1 . Without the flag, you get the pre-refactor pipeline; with it, you get the refactored one. The refactor has 3 stages: A, B, C.

1. Before Stage A

The following must be done before we officially kick off the refactor:

- Video MMU @Jayon (including fixing the current thinker max sequence parameter, etc.)
- Concurrency for Qwen3 Omni @xuesong
- Calibration threshold skills @yifei

Meanwhile, some tasks can run in parallel and make sense to finish either before or after the refactor. If a project is already underway (e.g., Ming's TP), or its release date falls before the refactor ends (e.g., Boson TTS), continue building it against the v0 architecture and add CI for v0.

Otherwise, the project should be paused and resumed after Stage A completes — e.g., Piecewise CUDA Graph @xuesong @jingwen and colocated placement @ratish.

Exit criteria for Stage A: the first alternative pipeline is merged in — i.e., real files are written into  sglang-omni/v1 . We're fully satisfied with the new pipeline's quality, and with  --use-v1  enabled, it passes all of v0's CI.

2. Stage A → B

The first alternative pipeline is done and can be activated via  --use-v1 . In this phase, migrate all features currently not protected by CI over to v1, and set up new CI that only guards v1.

Exit criteria for Stage B: v1 and v0 are fully equivalent.

3. Stage B → C

Profile, documentation, and blog post. Delete v0 and the  --use-v1  flag. Then we're cool.

**Comment by JingwenGu0829** [comment]
moved to #334
